oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/measure/BanyanDBTopologyQueryDAO.java

```diff
@@ -74,43 +78,34 @@ public List<Call.CallDetail> loadServiceRelationDetectedAtClientSide(Duration du
             throw new UnexpectedException("Service id is empty");
         }
 
-        List<QueryBuilder<MeasureQuery>> queryBuilderList = buildServiceRelationsQueries(serviceIds);
+        QueryBuilder<MeasureQuery> queryBuilder = buildServiceRelationsQuery(serviceIds);
 
-        return queryServiceRelation(duration, queryBuilderList, DetectPoint.CLIENT);
+        return queryServiceRelation(duration, queryBuilder, DetectPoint.CLIENT);
     }
 
     @Override
     public List<Call.CallDetail> loadServiceRelationsDetectedAtServerSide(Duration duration) throws IOException {
-        return queryServiceRelation(duration, Collections.singletonList(emptyMeasureQuery()), DetectPoint.SERVER);
+        return queryServiceRelation(duration, emptyMeasureQuery(), DetectPoint.SERVER);
     }
 
     @Override
     public List<Call.CallDetail> loadServiceRelationDetectedAtClientSide(Duration duration) throws IOException {
-        return queryServiceRelation(duration, Collections.singletonList(emptyMeasureQuery()), DetectPoint.CLIENT);
+        return queryServiceRelation(duration, emptyMeasureQuery(), DetectPoint.CLIENT);
     }
 
-    private List<QueryBuilder<MeasureQuery>> buildServiceRelationsQueries(List<String> serviceIds) {
-        List<QueryBuilder<MeasureQuery>> queryBuilderList = new ArrayList<>(serviceIds.size());
-        for (final String serviceId : serviceIds) {
-            queryBuilderList.add(new QueryBuilder<MeasureQuery>() {
-                @Override
-                protected void apply(MeasureQuery query) {
-                    query.and(eq(ServiceRelationServerSideMetrics.SOURCE_SERVICE_ID, serviceId));
-                }
-            });
-
-            queryBuilderList.add(new QueryBuilder<MeasureQuery>() {
-                @Override
-                protected void apply(MeasureQuery query) {
-                    query.and(eq(ServiceRelationServerSideMetrics.DEST_SERVICE_ID, serviceId));
-                }
-            });
-        }
-        return queryBuilderList;
+    private QueryBuilder<MeasureQuery> buildServiceRelationsQuery(List<String> serviceIds) {
+        return new QueryBuilder<MeasureQuery>() {
+            @Override
+            protected void apply(MeasureQuery query) {
+                query.or(in(ServiceRelationServerSideMetrics.SOURCE_SERVICE_ID, serviceIds))
+                        .or(in(ServiceRelationServerSideMetrics.DEST_SERVICE_ID, serviceIds));
+                query.groupBy(Sets.newHashSet(Metrics.ENTITY_ID, ServiceRelationServerSideMetrics.COMPONENT_IDS));
```

Comment:
The same as the comment left at line:311

Reply:
component id exists in service topology only, now.

Reply:
Fixed

Reply:
Please use `@SeriesID` to mark `entity_id` and `component_id`

Reply:
Could you upload the service_relation_server_side and service_relation_client_side measure's schema here for double-checking?

Reply:
They don't have `SeriesID` for now, please add them. `component_id` don't have to be a part of `SeriesID`, because they have chances to update when service IDs are not changed. 

@hanahmily Is that OK to keep it out of `SeriesID` but query them out as `group by`.

Reply:
I've added missing `@Series` to only `entity_id`.

Reply:
> @hanahmily Is that OK to keep it out of `SeriesID` but query them out as `group by`.

Yes. It is possible. But `groupBy` with entityID(s) save much memory according to the current implementation. If groupBy keys are not entityIDs, then hash-based physical plan will be used. Probably we may still need `componentIDs` as a part of the entity? @hanahmily

Reply:
> For update, this could be `ServA->ServB with component 1`, then update to `ServA->ServB with component 1,2,3`. 
> 
> This is the same entity from OAP perspective.

`GroupBy` will output two items in this scenario if the components ids is a part of the group key, whether it belongs to the seriesID or not. In other words, marking component ids as a part of seriesID affects the performance, not the semantics. 

If this is the case, I don't see any side effects to adding component ids to seriesID.

Reply:
Annotations added